hpy/debug/src/debug_handles.c

```diff
@@ -1,25 +1,98 @@
+        h = h->next;
+    }
+    h = info->closed_handles.head;
+    while(h != NULL) {
+        assert(h->is_closed);
+        h = h->next;
+    }
+#endif
+}
+
 DHPy DHPy_wrap(HPyContext *dctx, UHPy uh)
 {
     UHPy_sanity_check(uh);
     if (HPy_IsNull(uh))
         return HPy_NULL;
     HPyDebugInfo *info = get_info(dctx);
-    DebugHandle *handle = malloc(sizeof(DebugHandle));
-    if (handle == NULL) {
-        return HPyErr_NoMemory(info->uctx);
+
+    // if the closed_handles queue is full, let's reuse one of those. Else,
+    // malloc a new one
+    DebugHandle *handle = NULL;
+    if (info->closed_handles.size >= info->closed_handles_queue_max_size) {
+        handle = DHQueue_popfront(&info->closed_handles);
+    }
+    else {
+        handle = malloc(sizeof(DebugHandle));
+        if (handle == NULL) {
+            return HPyErr_NoMemory(info->uctx);
+        }
     }
     handle->uh = uh;
     handle->generation = info->current_generation;
-    handle->prev = NULL;
-    handle->next = info->open_handles;
-    if (info->open_handles)
-        info->open_handles->prev = handle;
-    info->open_handles = handle;
+    handle->is_closed = 0;
+    DHQueue_append(&info->open_handles, handle);
+    debug_handles_sanity_check(info);
     return as_DHPy(handle);
 }
 
+static void print_error(HPyContext *uctx, const char *message)
+{
+    // We don't have a way to propagate exceptions from within DHPy_unwrap, so
+    // we just print the exception to stderr and clear it
+    // XXX: we should use HPySys_WriteStderr when we have it
+    fprintf(stderr, "%s\n", message);
+    //HPyErr_PrintEx(0); // uncommment when we have it
+}
+
+
+// this is called when we try to use a closed handle
+void DHPy_invalid_handle(HPyContext *dctx, DHPy dh)
+{
+    HPyDebugInfo *info = get_info(dctx);
+    HPyContext *uctx = info->uctx;
+    DebugHandle *handle = as_DebugHandle(dh);
+    assert(handle->is_closed);
+    if (HPy_IsNull(info->uh_on_invalid_handle)) {
+        // default behavior: print an error and abort
+        HPy_FatalError(uctx, "Invalid usage of already closed handle");
+    }
+    /* call the custom callback but do NOT abort the execution. This
+       is useful e.g. on CPython where "closed handles" are still
+       actually valid until the refcount > 0: it should make it easier
+       to port extensions to HPy, e.g. by printing a warning inside
+       the callback and let the execution to continue, so that people
+       can fix the warnings one by one.
+    */
+    // XXX: so far the only call API that we have is CallTupleDict, we
+    // should use a more C-friendly variant as soon as we have it.
+    UHPy uh_args = HPy_NULL;
+    UHPy uh_res = HPy_NULL;
+    uh_args = HPyTuple_Pack(uctx, 0);
+    if (HPy_IsNull(uh_args)) {
+        print_error(uctx, "Error when preparing args to call the on_invalid_handle callback");
+        goto exit;
+    }
+    uh_res = HPy_CallTupleDict(uctx, info->uh_on_invalid_handle, uh_args, HPy_NULL);
```

Comment:
If `uh_args` is intended to be empty, then you should be able to pass `HPy_NULL` instead of making a tuple. At least that was my intention when I wrote `HPy_CallTupleDict`.

Reply:
good catch, thank you. This simplifies the code a lot